### ISSUE TITLE
Mac OS X Core Audio Layout problem

### ISSUE THREAD

**Issue by mpsrig** [mpv-player/mpv#3065]
Hello,

The handling of Core Audio layouts has a problem. 

On multichannel systems there are 2 layouts, the "stereo" layout and the "multichannel" layout.  The stereo layout is handled in an interesting way - it is "parsed" into an mpv layout the same way as the multichannel layout.  This should not be the case.  Rather, mpv should simply assume that the stereo layout is in fact stereo, and send left and right audio to those channels. 

Attached is an example of this on my system.  Left and Right are channels 11 and 12 on the audio interface, so mpv "parses" stereo as "sr-tc", has no idea what to do with this, and falls back to mono.

I realize that this may also be a symptom of a small bug in OS X but we can definitely work around this.  

`[ao] Trying audio driver 'coreaudio'
[ao/coreaudio] requested format: 48000 Hz, stereo channels, floatp
[ao/coreaudio] selected audio output device: Scarlett 8i6 USB (38)
[ao/coreaudio] input channel layout:
[ao/coreaudio] layout: tag: <0>, bitmap: <0>, descriptions <12>
[ao/coreaudio]  - description 0: label <4294967295, 64>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 1: label <4294967295, 64>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 2: label <4294967295, 64>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 3: label <4294967295, 64>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 4: label <4294967295, 64>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 5: label <4294967295, 64>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 6: label <4294967295, 64>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 7: label <4294967295, 64>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 8: label <4294967295, 64>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 9: label <4294967295, 64>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 10: label <1, 0>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 11: label <2, 1>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio] mp chmap: na-na-na-na-na-na-na-na-na-na-fl-fr
[ao/coreaudio] input channel layout:
[ao/coreaudio] layout: tag: <0>, bitmap: <0>, descriptions <2>
[ao/coreaudio]  - description 0: label <11, 10>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 1: label <12, 11>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio] mp chmap: sr-tc
[ao/coreaudio] Channel layouts:
[ao/coreaudio]  - na-na-na-na-na-na-na-na-na-na-fl-fr
[ao/coreaudio]  - sr-tc
[ao/coreaudio]  - mono
[ao/coreaudio] result: mono
[ao/coreaudio] Latency property cntl: 48 frames
[ao/coreaudio] Latency property zisf: 512 frames
[ao/coreaudio] Latency property tfas: 49 frames
[ao/coreaudio] audiounit latency [us]: 0
[ao/coreaudio] device latency [us]: 12687
[ao/coreaudio] using soft-buffer of 9600 samples.`

**Comment by wm4** [comment]
Yeah, we could just remove the code that reads and applies the stereo layout, and assume stereo is always supported.

But what's the point of the stereo layout at all?

**Comment by wm4** [comment]
CC @pigoz

**Comment by pigoz** [comment]
@mpsrig can you post a screenshot of how your speaker layout is configured in `Audio MIDI Setup.app`. This looks like misconfiguration on your end?

**Comment by mpsrig** [comment]
The point of stereo layout is to route stereo audio.  This can be separate from multichannel audio.  I.E. unless the app is sending surround sound, it should simply send stereo or mono audio to the system using the "stereo" path.  MPV does this fine but only in the mono case, not the stereo case.  Screenshots to follow.

**Comment by mpsrig** [comment]
![screenshot 2016-04-19 11 51 11](https://cloud.githubusercontent.com/assets/11206336/14645860/4fc17a5c-0625-11e6-943c-4d6f98d158ec.png)

**Comment by mpsrig** [comment]
![screenshot 2016-04-19 11 51 16](https://cloud.githubusercontent.com/assets/11206336/14645864/55630ee4-0625-11e6-839a-d4b49a4a7eae.png)

**Comment by pigoz** [comment]
Your config looks indeed correct, this looks like a bug in mpv.

**Comment by mpsrig** [comment]
In every other app this works.  All audio is stereo and is routed to 11 as left and 12 as right.  This is not a bug on my system.

**Comment by mpsrig** [comment]
Ideally what should happen is that unless the app is sending multichannel audio i.e. 5.1 or 7.1, it should drop to the stereo path.  I.E.  I should be able to have surround set up on 5.1 on interface channels 1-6 and stereo set up on 11-12 and it should work fine.  This works great in VLC.  Shouldn't be too hard a fix.

**Comment by mpsrig** [comment]
Update:  It seems that MPV has trouble with pretty much any nonstandard audio layout.  I changed my setup to be 5.1 on standard channels 1-6 and it still gets thrown off by channels 7-12 being "na" and outputs either mono (when the "stereo" output setup in Audio MIDI setup is set to 11 and 12) or stereo (when the "stereo" output is set to 1 and 2).  The input file is 5.1 and VLC does this fine.

**Comment by wm4** [comment]
Please post logs with -v for all interesting cases.

**Comment by mpsrig** [comment]
[case1-stereo_is11and12-multichannel_is_stereo11and12.txt](https://github.com/mpv-player/mpv/files/226610/case1-stereo_is11and12-multichannel_is_stereo11and12.txt)
[case2-standard-stereo1and2-multichannel_5.1-1thru6.txt](https://github.com/mpv-player/mpv/files/226609/case2-standard-stereo1and2-multichannel_5.1-1thru6.txt)
[case3-stereo11and12-multichannel_5.1-1thru6.txt](https://github.com/mpv-player/mpv/files/226608/case3-stereo11and12-multichannel_5.1-1thru6.txt)

**Comment by wm4** [comment]
And you never actually want 5.1 output? (Since a while ago we've switched back to stereo/mono downmix output by default.)

**Comment by mpsrig** [comment]
No, in the case where multichannel out is set to 5.1 on channels 1-6 I would expect it to do 5.1 output from a 5.1 source mkv - I guess I will have to look into the command line flags for this?

Michael Stella

**Comment by wm4** [comment]
Yes, you need to set ``--audio-channels=auto`` in this case.

**Comment by mpsrig** [comment]
Thanks for the tip.  There is obviously still a bug here though.  Will update.

**Comment by hmedia1** [comment]
Any advice for this scenario ?

In this case using `--ad-lavc-downmix=no --audio-device='coreaudio/SoundflowerEngine:1'`

```
[ao/coreaudio] layout has too many descriptions (64, max: 16)
[ao/coreaudio] input channel layout:
[ao/coreaudio] layout: tag: <0>, bitmap: <0>, descriptions <2>
[ao/coreaudio]  - description 0: label <1, 0>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio]  - description 1: label <2, 1>,  flags: <0>, coords: <0.000000, 0.000000, 0.000000>
[ao/coreaudio] mp chmap: stereo
[ao/coreaudio] Channel layouts:
[ao/coreaudio]  - stereo
[ao/coreaudio]  - mono
[ao/coreaudio] result: stereo
[ao/coreaudio] Latency property cntl: 0 frames
[ao/coreaudio] Latency property zisf: 512 frames
[ao/coreaudio] Latency property tfas: 0 frames
[ao/coreaudio] audiounit latency [us]: 333
[ao/coreaudio] device latency [us]: 10666
[ao/coreaudio] using soft-buffer of 9600 samples.
[cplayer] AO: [coreaudio] 48000Hz stereo 2ch floatp
[cplayer] AO: Description: CoreAudio AudioUnit
[af] Adding filter lavrresample 
[lavrresample] Remix: fl-fr-fc-lfe-bl-br-flc-frc-bc-sl-sr-tc-tfl-tfc-tfr-tbl -> stereo
[af] Audio filter chain:
[af]   [in] 48000Hz unknown16 (empty) 16ch floatp
[af]   [lavrresample] 48000Hz stereo 2ch floatp [a]
[af]   [out] 48000Hz stereo 2ch floatp
[af]   [ao] 48000Hz stereo 2ch floatp
```

Same happens if I use a different CoreAudio device which has 32 outputs.

I just want to use the first 16 of the outputs.

Using `--ao=pcm` writes the output to a file, but cannot get it to output on the first 16 channels of a CoreAudio interface that has a maximum of more than 16 channels.

**Comment by hmedia1** [comment]
> And you never actually want 5.1 output? (Since a while ago we've switched back to stereo/mono downmix output by default.)

How do we switch back to not doing this by default ?  It seems no matter what I do there's always a stereo remix happening.

**Comment by hmedia1** [comment]
Having a similar issue - added findings to #5042

**Comment by hmedia1** [comment]
> And you never actually want 5.1 output? (Since a while ago we've switched back to stereo/mono downmix output by default.)

Hi, what you're referring to here - the default downmix - where is this implemented ?  I am wondering if it plays a part in #5042
